packages/liveblocks-core/src/fsm.ts

```diff
@@ -0,0 +1,493 @@
+    return this;
+  }
+
+  /**
+   * Stops the state machine. Stopping the state machine will call exit
+   * handlers for the current state, but not enter a new state.
+   */
+  public stop(): void {
+    if (this.runningState !== RunningState.STARTED) {
+      throw new Error("Cannot stop a state machine that isn't started yet");
+    }
+    this.runningState = RunningState.STOPPED;
+    this.exit(null);
+    this.currentStateOrNull = null;
+  }
+
+  constructor(initialContext: Readonly<TContext>) {
+    this.runningState = RunningState.NOT_STARTED_YET;
+    this.currentStateOrNull = null;
+    this.states = new Set();
+    this.enterFns = new Map();
+    this.cleanupStack = [];
+    this.knownEventTypes = new Set();
+    this.allowedTransitions = new Map();
+    this.currentContext = Object.assign({}, initialContext);
+  }
+
+  public get context(): Readonly<TContext> {
+    return this.currentContext;
+  }
+
+  /**
+   * Define an explicit finite state in the state machine.
+   */
+  public addState(state: TState): this {
+    if (this.runningState !== RunningState.NOT_STARTED_YET) {
+      throw new Error("Already started");
+    }
+    this.states.add(state);
+    return this;
+  }
+
+  public onEnter(
+    nameOrPattern: TState | Wildcard<TState>,
+    enterFn: EnterFn<TContext>
+  ): this {
+    if (this.runningState !== RunningState.NOT_STARTED_YET) {
+      throw new Error("Already started");
+    } else if (this.enterFns.has(nameOrPattern)) {
+      throw new Error(
+        // TODO We _currently_ don't support multiple .onEnters() for the same
+        // state, but this is not a fundamental limitation. Just not
+        // implemented yet. If we wanted to, we could make this an array.
+        `enter/exit function for ${nameOrPattern} already exists`
+      );
+    }
+
+    this.enterFns.set(nameOrPattern, enterFn);
+    return this;
+  }
+
+  public onEnterAsync<T>(
+    nameOrPattern: TState | Wildcard<TState>,
+    promiseFn: (context: Readonly<TContext>) => Promise<T>,
+    onOK: Target<TContext, AsyncOKEvent<T>, TState>,
+    onError: Target<TContext, AsyncErrorEvent, TState>
+  ): this {
+    return this.onEnter(nameOrPattern, () => {
+      let cancelled = false;
+
+      void promiseFn(this.currentContext).then(
+        // On OK
+        (result: T) => {
+          if (!cancelled) {
+            this.transition({ type: "ASYNC_OK", result }, onOK);
+          }
+        },
+
+        // On Error
+        (error: unknown) => {
+          if (!cancelled) {
+            this.transition({ type: "ASYNC_ERROR", error }, onError);
+          }
+        }
+      );
+
+      return () => {
+        cancelled = true;
+      };
+    });
+  }
+
+  private getMatches(nameOrPattern: TState | Wildcard<TState>): TState[] {
+    const matches: TState[] = [];
+
+    // We're trying to match a group pattern here, i.e. `foo.*` (which might
+    // match `foo.bar` and `foo.qux` states)
+    if (nameOrPattern.endsWith("*")) {
```

Comment:
can you do `foo.bar*`? If not, maybe split on `.` and take the last index to make sure it's entirely "*".

Reply:
Fixed that by including the `.` in the check in 0fd123e383a13c1a263f075ba5db6aefd4578c93. Does that look okay to you? I wanted to avoid the unnecessary string split if we don't use the parts here.